From 8e54eeed30345ef36be38eb92ee3ad65d49d1cb9 Mon Sep 17 00:00:00 2001
From: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Date: Wed, 31 Jan 2018 11:20:01 +0200
Subject: [PATCH 3294/5242] MLK-15033-1: ASoC: fsl: Change rate constraints in
 TDM mode for AK4458

commit  0852455a37052e638334c2cfd86a96387f058177 from
https://source.codeaurora.org/external/imx/linux-imx.git

When in TDM mode, change constraints for rate and allow only
rates in [8KHz, 96KHz] due to the limitations of SAI master
clock. If rate is higher than 96KHz, the TX rate cannot be
obtained using only a 49MHz SAI clock.

Signed-off-by: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Reviewed-by: Daniel Baluta <daniel.baluta@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-ak4458.c |   23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/sound/soc/fsl/imx-ak4458.c b/sound/soc/fsl/imx-ak4458.c
index 30b649c..6b86777 100644
--- a/sound/soc/fsl/imx-ak4458.c
+++ b/sound/soc/fsl/imx-ak4458.c
@@ -38,6 +38,11 @@ struct imx_ak4458_data {
 	384000, 768000,
 };
 
+static const u32 ak4458_rates_tdm[] = {
+	8000, 16000, 32000,
+	48000, 96000,
+};
+
 static const u32 ak4458_channels[] = {
 	1, 2, 4, 6, 8, 10, 12, 14, 16,
 };
@@ -122,21 +127,16 @@ static int imx_aif_startup(struct snd_pcm_substream *substream)
 	static struct snd_pcm_hw_constraint_list constraint_channels;
 	int ret;
 
-	constraint_rates.list = ak4458_rates;
-	constraint_rates.count = ARRAY_SIZE(ak4458_rates);
-
-	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE,
-						&constraint_rates);
-	if (ret)
-		return ret;
-
-
 	if (data->tdm_mode) {
 		constraint_channels.list = ak4458_channels_tdm;
 		constraint_channels.count = ARRAY_SIZE(ak4458_channels_tdm);
+		constraint_rates.list = ak4458_rates_tdm;
+		constraint_rates.count = ARRAY_SIZE(ak4458_rates_tdm);
 	} else {
 		constraint_channels.list = ak4458_channels;
 		constraint_channels.count = ARRAY_SIZE(ak4458_channels);
+		constraint_rates.list = ak4458_rates;
+		constraint_rates.count = ARRAY_SIZE(ak4458_rates);
 	}
 
 	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_CHANNELS,
@@ -144,6 +144,11 @@ static int imx_aif_startup(struct snd_pcm_substream *substream)
 	if (ret)
 		return ret;
 
+	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE,
+						&constraint_rates);
+	if (ret)
+		return ret;
+
 	return 0;
 }
 
-- 
1.7.9.5

